test/jdk/java/nio/jni/NewDirectByteBuffer.java

```diff
@@ -0,0 +1,143 @@
+        assertTrue(buf.isDirect(), "Buffer is not direct");
+        assertFalse(buf.hasArray(), "Buffer has an array");
+        if (capacity > 0) {
+            assertTrue(buf.hasRemaining(), "Buffer has no remaining values");
+        }
+        assertFalse(buf.isReadOnly(), "Buffer s read-only");
+        assertEquals(capacity, buf.capacity(),
+            "Buffer::capacity returned unexpected value");
+        assertEquals(0L, buf.position(),
+            "Buffer::position returned unexpected value");
+        assertEquals(capacity, buf.limit(),
+            "Buffer::limit returned unexpected value");
+
+        // Verify that the various state mutators work correctly
+        int halfPos = buf.capacity()/2;
+        buf.position(halfPos);
+        assertEquals(halfPos, buf.position(),
+            "Position not set to halfPos");
+        assertEquals(buf.capacity() - halfPos, buf.remaining(),
+            "Remaining not capacity - halfPos");
+
+        buf.mark();
+
+        int twoThirdsPos = 2*(buf.capacity()/3);
+        buf.position(twoThirdsPos);
+        assertEquals(twoThirdsPos, buf.position(),
+            "Position not set to twoThirdsPos");
+        assertEquals(buf.capacity() - twoThirdsPos, buf.remaining(),
+            "Remaining != capacity - twoThirdsPos");
+
+        buf.reset();
+        assertEquals(halfPos, buf.position(),
+            "Buffer not reset to halfPos");
+
+        buf.limit(twoThirdsPos);
+        assertEquals(twoThirdsPos, buf.limit(),
+            "Limit not set to twoThirdsPos");
+        assertEquals(twoThirdsPos - halfPos, buf.remaining(),
+            "Remaining != twoThirdsPos - halfPos");
+
+        buf.position(twoThirdsPos);
+        assertFalse(buf.hasRemaining(), "Buffer has remaining values");
+    }
+
+    @ParameterizedTest
+    @ValueSource(longs = {0L, 1L, (long)Integer.MAX_VALUE/2,
+        (long)Integer.MAX_VALUE - 1, (long)Integer.MAX_VALUE})
+    void legalCapacities(long capacity) {
+        long addr;
+        try {
+            addr = UNSAFE.allocateMemory(capacity);
+        } catch (OutOfMemoryError ignore) {
+            System.err.println("legalCapacities( " + capacity
+                + ") test skipped due to insufficient memory");
+            return;
+        }
+        try {
+            ByteBuffer buf = newDirectByteBuffer(addr, capacity);
+            assertEquals(addr, getDirectBufferAddress(buf),
+                "GetDirectBufferAddress does not return supplied address");
+            checkBuffer(buf, capacity);
+        } finally {
+            UNSAFE.freeMemory(addr);
+        }
+    }
+
+    @ParameterizedTest
+    @ValueSource(longs = {Long.MIN_VALUE, (long)Integer.MIN_VALUE - 1L, -1L,
+        (long)Integer.MAX_VALUE + 1L, 3_000_000_000L, 5_000_000_000L,
+        Long.MAX_VALUE})
+    void illegalCapacities(long capacity) {
+        assertThrows(IllegalArgumentException.class, () -> {
+            long addr = UNSAFE.allocateMemory(capacity);
```

Comment:
How is `allocateMemory` going to respond to these illegal capacity requests?

If we expect `newDirectByteBuffer` to throw then can't we use a dummy buffer value as it will never be used?

Reply:
> How is `allocateMemory` going to respond to these illegal capacity requests?
> 
> If we expect `newDirectByteBuffer` to throw then can't we use a dummy buffer value as it will never be used?

For illegalCapacities it should call allocateMemory with some fixed value, say 1, not capacity. Surprised the allocateMemory isn't failing for the test inputs.

Reply:
Changed allocation size to `1` in `illegalCapacities` in 3b1eef25b53dbeb36e4b3a8d6c279a5e0f5a5ae3.